Reduce allocations in findMember

diesalbla opened the pull request:
This commit tries to reduce memory allocations noticed in the issue https://github.com/scala/scala-dev/issues/493 by @retronym.

#### Possible Cause of the Excessive Allocations

The following line:
```Scala
refinementParents :::= currentBaseClass.parentSymbols
```
Is pre-prending the list `currentBaseClass.parentSymbols` to the previous `refinementParens`. Now, suppose `initBaseClasses` contains `n` symbols for which `initBaseClasses` is true, and suppose that each of those symbols has `m` elements in its `parentSymbols`. You would then end up with a list of length `m * n`, which would have that many cons (`::`) objects. Moreover, each call to `parentSymbols` is not to a `val`, but a `def`, which means that it creates a list of length `m`, and then has to copy that list (to prepend is to copy) and throw that away. 

####  Fix

Instead of concatenating, we keep a list of lists, one for each refinement class. This could, ideally, halve the memory leak due to that line. For the previous example, it will reduce the number of `Cons` object created from `2 * m * n` to `n + n * m`. This will be more effective if those `parentSymbols` lists are longer.

retronym commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -110,7 +110,7 @@ trait FindMembers {
           //
           //           => private members should be included from T1, ... Tn. (scala/bug#7475)
-          refinementParents :::= currentBaseClass.parentSymbols
+          refinementParents ::= currentBaseClass.parentSymbols
```
My hope was to defer the call to `parentSymbols` until it was needed. So just add `currentBaseClass` to a list of refinement symbols, and then change `refinementParents.exists(_.contains(owner))` below to create `refinementParents ` on first access.

retronym commented:
Something along the lines of:

```diff
$ git diff head -- src/reflect/scala/reflect/internal/tpe/FindMembers.scala
diff --git a/src/reflect/scala/reflect/internal/tpe/FindMembers.scala b/src/reflect/scala/reflect/internal/tpe/FindMembers.scala
index aa56a9a6e3..b1a7e86842 100644
--- a/src/reflect/scala/reflect/internal/tpe/FindMembers.scala
+++ b/src/reflect/scala/reflect/internal/tpe/FindMembers.scala
@@ -15,6 +15,7 @@ trait FindMembers {
   /** Implementation of `Type#{findMember, findMembers}` */
   private[internal] abstract class FindMemberBase[T](tpe: Type, name: Name, excludedFlags: Long, requiredFlags: Long) {
     protected val initBaseClasses: List[Symbol] = tpe.baseClasses
+    private var refinementBaseClasses: List[Symbol] = Nil

     // The first base class, or the symbol of the ThisType
     // e.g in:
@@ -93,7 +94,7 @@ trait FindMembers {
           if (meetsRequirements) {
             val excl: Long = flags & excluded
             val isExcluded: Boolean = excl != 0L
-            if (!isExcluded && isPotentialMember(sym, flags, currentBaseClass, seenFirstNonRefinementClass, refinementParents)) {
+            if (!isExcluded && isPotentialMember(sym, flags, currentBaseClass, seenFirstNonRefinementClass)) {
               if (shortCircuit(sym)) return false
               else addMemberIfNew(sym)
             } else if (excl == DEFERRED) {
@@ -110,7 +111,7 @@ trait FindMembers {
           //           the component types T1, ..., Tn and the refinement {R }
           //
           //           => private members should be included from T1, ... Tn. (scala/bug#7475)
-          refinementParents :::= currentBaseClass.parentSymbols
+          refinementBaseClasses ::= currentBaseClass
         else if (currentBaseClass.isClass)
           seenFirstNonRefinementClass = true // only inherit privates of refinement parents after this point

@@ -130,7 +131,7 @@ trait FindMembers {
     // Q. When does a potential member fail to be an actual member?
     // A. if it is subsumed by an member in a subclass.
     private def isPotentialMember(sym: Symbol, flags: Long, owner: Symbol,
-                                  seenFirstNonRefinementClass: Boolean, refinementParents: List[Symbol]): Boolean = {
+                                  seenFirstNonRefinementClass: Boolean): Boolean = {
       // conservatively (performance wise) doing this with flags masks rather than `sym.isPrivate`
       // to avoid multiple calls to `Symbol#flags`.
       val isPrivate      = (flags & PRIVATE) == PRIVATE
@@ -142,7 +143,7 @@ trait FindMembers {
              !isPrivateLocal // private[this] only a member from within the selector class. (Optimization only? Does the spec back this up?)
           && (
                   !seenFirstNonRefinementClass
-               || refinementParents.contains(owner)
+               || refinementBaseClasses.exists(_.info.parents.exists(_.typeSymbol == owner))
              )
         )
```

diesalbla commented:
@retronym The second set of changes tries to address that. I think it includes some of the lines in your diff above. 

The reason why I use a custom data structure, rather than a standard collection, is that I could not think of another way to cover a scenario with alternating evaluated and non-evaluated list of parents. For instance, you could add three classes `c1,c2,c3`, then evaluate the parents of `c3` as `p3`, and only that (they are searched in LIFO order), and then add more classes `c4,c5,c6` whose lists of parents were never used.

retronym commented:
Typically, there are 0 refinement classes in the base class list.

Sometimes, there is 1 or 2.

```
scala> val t1 = if ("".isEmpty) List("") else Map("" -> "")
t1: scala.collection.immutable.Iterable[java.io.Serializable] with PartialFunction[Int with String,String]{def seq: scala.collection.immutable.Iterable[java.io.Serializable] with PartialFunction[Int with String,String]{def seq: scala.collection.immutable.Iterable[java.io.Serializable] with PartialFunction[Int with String,String]}} = List("")

scala> val bts = typeOf[t1.type].baseTypeSeq
bts: $r.intp.global.BaseTypeSeq = BTS(<refinement>.type,scala.collection.immutable.Iterable[java.io.Serializable] with PartialFunction[Int with String,String]{def seq: scala.collection.immutable.Iterable[java.io.Serializable] with PartialFunction[Int with String,String]{def seq: scala.collection.immutable.Iterable[java.io.Serializable] with PartialFunction[Int with String,String]}},scala.collection.immutable.Iterable[java.io.Serializable],Iterable[java.io.Serializable],scala.collection.immutable.Traversable[java.io.Serializable],Traversable[java.io.Serializable],scala.collection.IterableLike[java.io.Serializable,Iterable[java.io.Serializable]] with scala.collection.IterableLike[java.io.Serializable,scala.collection.immutable.Iterable[java.io.Serializable]],scal...
```

Esoteric case with two refinements in the base type sequence:

```
scala> trait C { self: (D with (G with F { def foo: Int })) => }; class D; class E; class G; class F
defined trait C
defined class D
defined class E
defined class G
defined class F

scala> typeOf[C].typeOfThis
res5: $r.intp.global.Type = C with D with G with F{def foo: Int}

scala> typeOf[C].typeOfThis.baseTypeSeq
res6: $r.intp.global.BaseTypeSeq = BTS(C with D with G with F{def foo: Int},D with G with F{def foo: Int},G with F{def foo: Int},C,D,G,F,Object,Any)

scala> typeOf[C].typeOfThis.baseClasses.map(_.isRefinementClass)
res7: List[Boolean] = List(true, true, true, false, false, false, false, false, false)
```

Even when we find a refinement class in the BTS, `findMember` typically doesn't need to check if the found member is owned by a parent of one of the refinements. This is only in the special case when the found member is `private`.

So the goal here is not to make `refinementParents.contains(owner)` fast, but instead to avoid calling `parentSymbols` before we need to.

diesalbla commented:
Well, I am afraid I did not know those numbers, but it is useful information. The solution, as is, does delay the call to `parentSymbols` until it is first needed, but with some overkill, and with the little overhead of two arrays that, from what you say, will be mostly empty. I will fix that.

diesalbla commented:
@retronym I have reimplemented the auxiliary class, to use a linked List of Either, instead of the two arrays. This solution should have a negligible overhead, in number of objects allocated, provided that 1) there are usually few refinement classes in the base class list, and 2) the need to perform the call to `parentSymbols` is very rare.

retronym commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -299,4 +298,50 @@ trait FindMembers {
+
+  /* A variable-length linked list, where each element is either a refinement class, 
+   * or the list of parents of that class. */
+  private[tpe] final class ClassesWithParents(length: Int) {
```
length is unused now.

I actually would prefer not to allocate `ClassesWithParents`, and roll the fields/logic here into `FindMemberBase`

retronym commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -299,4 +298,50 @@ trait FindMembers {
+
+    /* Returns true if any of the classes has the cl class as a parent.
+     * In the process, it reads the list of parents for those classes for which we had not fetched them.
+     * The list of classes are searched in Last-In-First-Out order.
```
The order of search doesn't matter AFAICT, which might make the implementation simpler.

retronym commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -59,6 +59,29 @@ trait FindMembers {
+            val pp = hh.parentSymbols
+            refinedParents ::= pp
+            found = pp.contains(cl)
+          case Nil => // should not occur
```
I think `isParentOfAnyRefinementClass ` would be simpler, and just as quick, if replaced with:

`refinedClasses.exists(_.info.parents.exists(_.typeSymbol == cl)))`

This basically inlines `def parentSymbols = info.parents.map(_.typeSymbol)`, avoiding the list creation.

The two closures in that expression should be compiled to be statically hoisted instances by `LambdaMetafactory`, so that doesn't add allocations.

retronym commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -142,11 +161,11 @@ trait FindMembers {
         )
 
-      (!isPrivate || admitPrivate(sym)) && (sym.name != nme.CONSTRUCTOR || owner == initBaseClasses.head)
+      (sym.name != nme.CONSTRUCTOR || owner == initBaseClasses.head) && (!isPrivate || (selectorClass == owner) || admitPrivate(sym))
```
What was the reason for changing the order of the checks here?

diesalbla replied:
The reason was to make use of the short-circuiting of boolean operators, so as to delay the evaluation of the call to `admitPrivate` method until it is strictly needed.

diesalbla commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -130,23 +131,19 @@ trait FindMembers {
       val isPrivateLocal = (flags & PrivateLocal) == PrivateLocal
 
       // TODO Is the special handling of `private[this]` vs `private` backed up by the spec?
-      def admitPrivate(sym: Symbol): Boolean =
```
As it turns out, the `sym` parameter was not being used.

retronym commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -110,7 +111,7 @@ trait FindMembers {
           //
           //           => private members should be included from T1, ... Tn. (scala/bug#7475)
-          refinementParents :::= currentBaseClass.parentSymbols
+          refinedClasses ::= currentBaseClass
```
`walkBaseClasses` is actually called twice from `searchConcreteThenDeferred`, so we'll end up accumulating elements twice in the list of `refinedClasses`. Perhaps it is better to make it a local variable than a field.

retronym committed:
Adds benchmarks for findMember.

- Adds the new benchmarks subproject to the IntelliJ config
- Add a benchmark for findMember

retronym commented:
Once the benchmark is setup, you can also use the Flight Recorder profiler provided by sbt-jmh

```
$ sbt setupPublishCore 'bench/jmh:run scala.reflect.internal.FindMemberBenchmark.findMemberRefined -prof jmh.extras.JFR -f1'
.. 
[info] Flight Recording output saved to:
[info]   /Users/jz/code/scala/test/benchmarks/./scala.reflect.internal.FindMemberBenchmark.findMemberRefined-Throughput-0.jfr
```

That shows that all the remaining allocation is of the `FindMember` class itself (I've prototyped a way to get rid of that, too, in https://github.com/scala/scala-dev/issues/498)

diesalbla committed:
Fixes Issue 493, to avoid unnecessary allocation.

This commit avoids unnecessary memory use noticed
in Issue 493 of https://github.com/scala/scala-dev.

The line: `refinementParents :::= currentBaseClass.parentSymbols`
in each iteration pre-prends `currentBaseClass.parentSymbols` to the
previous `refinementParens`. Thus, at the end of the loop,
the length of `refinementParens` is the sum of the `parentSymbols`
lists obtained from each symbol in `initBaseClasses` which is a
refinement classes. That creates as many cons (`::`) objects.
Moreover, since `parentSymbols` is not a `val` but a `def`, it creates
a list of length `m`, copies it (to prepend it) and the throws it away.

To prevent these allocations, we replace the flattened
`refinedParents` list by a `refinedClasses` list, which
stores the classes whose parents we have not yet looked into.
We just use the `exists` methods of the List class to look for a
refinedClass with at least one parent is the currentBaseClass.

retronym commented:
Thanks for your contribution! I think the result we is both faster _and_ simpler than before :)

som-snytt commented:
Thanks, this was an interesting read. I only saw it because of my github feed, which I appreciate.
